### DISCUSSION TITLE
Discussion: Extending Cloud Providers

### DISCUSSION THREAD

**Discussion by jonathan-innis** [kubernetes-sigs/karpenter#2820] [General]
### Tell us about your request

There's a growing number of requests to extend the cloud providers that Karpenter supports from AWS to other cloud providers (Azure, GCP, Orcale, etc.). This issue is intended to cover the discussion of extending the number of cloud providers broadly.

### Additional Context

__Related Issues__

aws/karpenter#2507 
aws/karpenter#936

__Tracking Tasks__

- [ ] #754

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ellistarn** [comment]
I'd like to explore two models for this:
1: Karpenter as a Library (as currently implemented here: https://github.com/bwagner5/karpenter-k3d)
2: Karpenter as an API: https://github.com/kubernetes-csi

A few table stakes in my mind:
1. Vendors should be able to define their own API surface (e.g. AWSNodeTemplate) that's separately versioned
2. Vendor code should live in a separate repository from Karpenter core (k8s is doing this across the board w/ in-tree / out-of-tree)

**Comment by debugger24** [comment]
Happy to work on this. Let me know how we can collaborate.

**Comment by jonathan-innis** [comment]
@debugger24 Are you interested in helping/building a particular provider? GCP? Oracle? Digital Ocean?

**Comment by tonymarkel** [comment]
I'd like to help with the Oracle portions of this in whatever capacity I can. If I read @ellistarn correctly, I'm okay to create a separate repo like @bwagner5 did. I haven't been able to find an equivalent open-source project on the 'net, so hopefully I'm not re-inventing a wheel. If I don't see anything in this thread in a day or so, I'll follow Brandon's example and put what little I have in a repo and share it here.

**Comment by jonathan-innis** [comment]
Nope! @tonymarkel. Feel free to take a stab at the Oracle cloudprovider and share the repo in this issue!

**Comment by ellistarn** [comment]
The best example you'll find is to look at aws/karpenter as the aws implementation. You'll need your provider to rely on aws/karpenter-core.

**Comment by sftim** [comment]
Should we label this with _neutrality_?

**Comment by Anhui-tqhuang** [comment]
seems that if we want to make a cloud provider, we cannot reuse the existing repo as it is tied to AWS implementation, we need to use karpenter-core to create our own karpenter ....

many implementation is not interfaced like
- https://github.com/aws/karpenter/tree/main/pkg/providers/instance
- https://github.com/aws/karpenter/tree/main/pkg/providers/instancetype

**Comment by ellistarn** [comment]
For an example of another provider implementation, check out: https://github.com/Azure/karpenter

**Comment by jackfrancis** [comment]
> For an example of another provider implementation, check out: https://github.com/Azure/karpenter

Note the source attribution section:

https://github.com/Azure/karpenter#source-attribution

Which is to say: for folks looking to rapidly prototype a new provider the AKS provider above is proof that you can just take the existing AWS bits and replace the relevant AWS-specific stuff with your own stuff.

cc @Anhui-tqhuang 

There are valid points about how we can improve the portability of the various provider interfaces, but it shouldn't be a blocker per se for folks who are eager to build now.

**Comment by jonathan-innis** [comment]
Looking over this issue: I'm realizing that what this issue really is is a discussion issue around a bunch of things, providing a common space for any cloud providers that want to implement a Karpenter cloud provider to come to have an active discussion about how to do it.

There's no true exit criteria for this issue, so I'm changing this to be prefixed with "Discussion:"

**Comment by sftim** [comment]
We can get this converted to a GitHub Discussion if you like, @jonathan-innis.

**Comment by jonathan-innis** [comment]
> We can get this converted to a GitHub Discussion

@sftim My one reservation about converting this to a discussion is routing. Do we have a good feeling which issues will get routed as discussions and which ones will just be bare-bones issues. I worry if we start having separate discussions for some one-off issues over in the Discussions tab in Github that we are going to lose people who are used to going to the issues section.

I also don't want this to be the _only_ issue that ends up converting into a Discussion. Realistically, I see a lot of "discussion-y" issues that probably still make sense classified as issues so I wonder if we just keep this as-is for now.

Happy to hear any additional insight that you might have here.

**Comment by Bryce-Soghigian** [comment]
Another Datapoint this thread might be interested in: https://github.com/kubernetes/autoscaler/issues/5394#issuecomment-1485331666

[a comment by an automated account is not shown]

**Comment by till** [comment]
I would argue that the outcome of this ticket is documentation that implements a dummy provider with docker containers or something simple. 😆

**Comment by tzneal** [comment]
There is a simple cloud provider at https://github.com/kubernetes-sigs/karpenter/tree/main/kwok that uses [KWOK](https://kwok.sigs.k8s.io/)

[a comment by an automated account is not shown]

**Comment by jpark0910** [comment]
Has there been any progress on the development for an Oracle Cloud provider for Karpenter?

**Comment by manibalaji-prod** [comment]
Looks like i need to go in-depth on the GoLang and start writing my own Karpenter.

**Comment by wy0824** [comment]
We have impl the karpenter for oracle cloud, and will open source ASAP 

> Has there been any progress on the development for an Oracle Cloud provider for Karpenter?

**Comment by gawbul** [comment]
Anyone aware of ongoing development of a GCP provider?

**Comment by dm3ch** [comment]
> Anyone aware of ongoing development of a GCP provider?

There's project of GCP provider but it's on very early stage (at the same time it's already. mentioned in Karpenter README.md) - https://github.com/cloudpilot-ai/karpenter-provider-gcp 

Unfortunately there's a very high chance that it wouldn't anyway work with GKE, cause it doesn't allow of joining own managed nodes.

In the mean time there's a chance that it's already possible to use Karpenter on GCP using Cluster API provider, but I haven't tested it yet

**Comment by jpedrosa94** [comment]
> We have impl the karpenter for oracle cloud, and will open source ASAP 
> 
> 
> 
> > Has there been any progress on the development for an Oracle Cloud provider for Karpenter?
> 
> 
> 
> 

Hey, do you have any updates regarding share OKE implementation?

**Comment by dsyoi** [comment]
I ( part of [Kubeify](https://kubeify.com/) team) started prototyping provider for Digitalocean cloud at https://github.com/k8ubeify/karpenter-provider-doks. 

> [!NOTE]
> The code is very pre-alpha without test-cases in `main`, we are steady making progress in feature branch.

**Comment by maximveksler** [comment]
> > Anyone aware of ongoing development of a GCP provider?
> 
> There's project of GCP provider but it's on very early stage (at the same time it's already. mentioned in Karpenter README.md) - https://github.com/cloudpilot-ai/karpenter-provider-gcp
> 
> Unfortunately there's a very high chance that it wouldn't anyway work with GKE, cause it doesn't allow of joining own managed nodes.
> 
> In the mean time there's a chance that it's already possible to use Karpenter on GCP using Cluster API provider, but I haven't tested it yet

The issue of self managed nodes on GKE seems to be solvable https://github.com/cloudpilot-ai/karpenter-provider-gcp/issues/7

**Comment by little-cat-lover** [comment]
> We have impl the karpenter for oracle cloud, and will open source ASAP
> 
> > Has there been any progress on the development for an Oracle Cloud provider for Karpenter?

Hi @wy0824 could you please share the progress of OSing it? I'm working on my implementation and it would be a great help.

**Comment by Joanale273** [comment]
is there any hope to deploy karpenter on Huawei Cloud?

**Comment by wy0824** [comment]
@little-cat-lover It will be open source in two weeks. Our company's open source process is slow.

**Comment by acarvalhopt** [comment]
> > We have impl the karpenter for oracle cloud, and will open source ASAP
> > > Has there been any progress on the development for an Oracle Cloud provider for Karpenter?
> 
> Hey, do you have any updates regarding share OKE implementation?

There is any update on this?

**Comment by wy0824** [comment]
> > > We have impl the karpenter for oracle cloud, and will open source ASAP
> > > > Has there been any progress on the development for an Oracle Cloud provider for Karpenter?
> > 
> > 
> > Hey, do you have any updates regarding share OKE implementation?
> 
> There is any update on this?

Yes, we are in the final step, will open it to the community this week.

**Comment by wy0824** [comment]
Hello everyone, we open-sourced our karpenter-oci project: https://github.com/zoom/karpenter-oci. Thanks for waiting, issues and PR are welcome. @acarvalhopt @little-cat-lover

**Comment by acarvalhopt** [comment]
Hello, thank you. We'll run some tests, and if we encounter any issues, I'll report them.

[a comment by an automated account is not shown]

**Comment by ShadowJonathan** [comment]
@wy0824 any updates?

**Comment by wy0824** [comment]
> [@wy0824](https://github.com/wy0824) any updates?

refer: https://github.com/zoom/karpenter-oci

**Comment by nicolas-g** [comment]
what about GCP ?

**Comment by dm3ch** [comment]
> what about GCP ?

https://github.com/cloudpilot-ai/karpenter-provider-gcp

[a comment by an automated account is not shown]

**Comment by ShadowJonathan** [comment]
@wy0824 is this issue intended to be closed?

**Comment by jmdeal** [comment]
I don't think this is really an actionable issue, but I do think it's valuable to have a place to discuss provider implementation. Of course there's the k8s slack but I think it's reasonable to keep this open as a discussion. I'll open it back up and transfer it over.

[a comment by an automated account is not shown]

**Comment by sophie-warner-annalise-ai** [comment]
Is there a chance that the other cloud major cloud provider's implementations would be merged into this one once they reach stability?
